File: 14889_스타트와 링크.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int arr[22][22];
int N, result = 9999999;
int aa[12], bb[12];

int calc()
{
	int a = 0, b = 0;

	for (size_t i = 0; i < N / 2; i++)
	{
		for (size_t j = 0; j < N / 2; j++)
		{
			a += (arr[aa[i]][aa[j]] + arr[aa[j]][aa[i]]);
			b += (arr[bb[i]][bb[j]] + arr[bb[j]][bb[i]]);
		}
	}
	return abs(a - b);
}

void func(int n, int a, int b)
{
	if (n == N)
	{
		result = min(result, calc());
		return;
	}

	if (a < N / 2)
	{
		aa[a] = n + 1;
		func(n + 1, a + 1, b);
		aa[a] = 0;
	}

	if (b < N / 2)
	{
		bb[b] = n + 1;
		func(n + 1, a, b + 1);
		bb[b] = 0;
	}

	

}

int main()
{

	cin >> N;
	for (size_t i = 1; i <= N; i++)
	{
		for (size_t j = 1; j <= N; j++)
		{
			cin >> arr[i][j];
		}
	}

	func(0, 0, 0);

	cout << result/2 << endl;




	return 0;
}

File: 삼성기출/14889_스타트와 링크.py
N = int(input())

arr = [list(map(int, input().split())) for _ in range(N)]
result = 20*19*100

def stat(team):
    answer = 0
    for i, num in enumerate(team):
        for j in range(i+1, len(team)):
            x, y = num, team[j]
            answer += arr[x][y] + arr[y][x]
    return answer

def func(n, a, b):
    global result
    if n == N:
        result = min(abs(stat(a)-stat(b)), result)
    else:
        if len(a) < N/2:
            a.append(n)
            func(n+1, a, b)
            a.remove(n)
        if len(b) < N/2:
            b.append(n)
            func(n+1, a, b)
            b.remove(n)
    return

func(0, [], [])
print(result)
